src/compiler/scala/tools/nsc/transform/patmat/Logic.scala

```diff
@@ -113,14 +113,19 @@ trait Logic extends Debugging {
       def implications: List[(Sym, List[Sym], List[Sym])]
     }
 
+    // The error message of t7020 assumes the ops are ordered implicitly
+    // However, ListSet is slow (concatenate cost?), so use
+    // java.util.LinkedHashSet (which grantees "the order in which elements were inserted into the set")
+    import scala.jdk.CollectionConverters._
+
     // would be nice to statically check whether a prop is equational or pure,
     // but that requires typing relations like And(x: Tx, y: Ty) : (if(Tx == PureProp && Ty == PureProp) PureProp else Prop)
     final case class And(ops: Set[Prop]) extends Prop
     object And {
       def apply(ps: Prop*)           = create(ps)
       def create(ps: Iterable[Prop]) = ps match {
         case ps: Set[Prop] => new And(ps)
-        case _             => new And(ps.to(scala.collection.immutable.ListSet))
+        case _             => new And(new java.util.LinkedHashSet[Prop](ps.toSet.asJava).asScala.to(Set))
```

Comment:
`ps.toSet` creates an immutable HashSet which has non-deterministic iteration order.

I think we can use Scala's `mutable.LinkedHashSet` instead of Java's. Maybe the best option is to change the `And` and `Or` case classes to have a `mutable.LinkedHashSet[Prop]` parameter? That would express the fact that an ordered collection is required.

Otherwise, use `collection.Set[Prop]`?

@dwijnand what do you think?

Reply:
Isn't `java.util.LinkedHashSet` known to be faster than `mutable.LinkedHashSet`?  Good catch on the toSet: we need to build up the set in insertion-order of the iterable.

I don't mind if we change the parameter either.

Reply:
The Java version is possibly still faster, but our LinkedHashSet was recently brought up to date (https://github.com/scala/scala/pull/10221). Compared to ListSet which we are replacing here, it's likely going to be fine.